File: ui/lib/tournament_ui/optimizer_runs.ex
defmodule TournamentUi.OptimizerRuns do
  @moduledoc """
  Read-only Elixir adapter over the Python-side `optimizer_run` table in the
  central judgement-fabric DB.

  The Python optimizer runner writes rows here (see bin/optimizer_runs.py);
  the LiveView reads from here so it can show "still running", "done", logs,
  and result candidate version after the user comes back later.
  """

  alias Exqlite.Sqlite3

  defp db_path do
    home = System.get_env("DATA_TOURNAMENTS_HOME") || "/tmp/data-tournaments"
    Path.join(home, "judgements.db")
  end

  @doc """
  Create a `running` row and return its id.

  Mirrors `bin.optimizer_runs.start()` so the LiveView can create the
  persistent record before spawning the script that updates it.
  """
  def start(opts) do
    domain = Keyword.fetch!(opts, :domain)
    target = Keyword.fetch!(opts, :target)
    rubric = Keyword.get(opts, :rubric)
    prompt_name = Keyword.get(opts, :prompt_name)

    sql =
      "INSERT INTO optimizer_run(domain, target, rubric, prompt_name) " <>
        "VALUES (?, ?, ?, ?)"

    path = db_path()

    with {:ok, conn} <- Sqlite3.open(path),
         {:ok, st} <- Sqlite3.prepare(conn, sql),
         :ok <- Sqlite3.bind(st, [domain, target, rubric, prompt_name]),
         :done <- Sqlite3.step(conn, st),
         {:ok, id} <- Sqlite3.last_insert_rowid(conn) do
      Sqlite3.release(conn, st)
      Sqlite3.close(conn)
      id
    end
  end

  @doc """
  Return all optimizer runs for a domain, newest first.

  Each row is a map with atom keys:
    %{
      id: integer,
      domain: string,
      target: string,                 # "judge" | "generator"
      status: string,                 # "running" | "done" | "error" | "canceled"
      log: string,                    # accumulated stdout lines
      result: map | nil,              # decoded JSON
      rubric: string | nil,
      prompt_name: string | nil,
      exit_code: integer | nil,
      started_at: string,
      finished_at: string | nil
    }
  """
  def list_for_domain(domain) when is_binary(domain) do
    sql =
      "SELECT id, domain, target, rubric, prompt_name, status, exit_code, " <>
        "log, result, started_at, finished_at " <>
        "FROM optimizer_run WHERE domain = ? ORDER BY id DESC"

    case query(sql, [domain]) do
      {:ok, rows} -> Enum.map(rows, &row_to_map/1)
      _ -> []
    end
  end

  @doc """
  Return the most recent optimizer run matching domain + target.

  Returns `nil` if no run exists. Useful for the "Optimize judge" /
  "Optimize generator" buttons which need to know the current state to
  decide whether to render "Run now" or "Running… (last started 2 min ago)".
  """
  def latest(opts) do
    domain = Keyword.fetch!(opts, :domain)
    target = Keyword.fetch!(opts, :target)

    sql =
      "SELECT id, domain, target, rubric, prompt_name, status, exit_code, " <>
        "log, result, started_at, finished_at " <>
        "FROM optimizer_run WHERE domain = ? AND target = ? " <>
        "ORDER BY id DESC LIMIT 1"

    case query(sql, [domain, target]) do
      {:ok, [row | _]} -> row_to_map(row)
      _ -> nil
    end
  end

  # ── private ────────────────────────────────────────────────────────────

  defp row_to_map([
         id,
         domain,
         target,
         rubric,
         prompt_name,
         status,
         exit_code,
         log,
         result_json,
         started_at,
         finished_at
       ]) do
    %{
      id: id,
      domain: domain,
      target: target,
      rubric: rubric,
      prompt_name: prompt_name,
      status: status,
      exit_code: exit_code,
      log: log || "",
      result: decode_json(result_json),
      started_at: started_at,
      finished_at: finished_at
    }
  end

  defp decode_json(nil), do: nil
  defp decode_json(""), do: nil

  defp decode_json(s) when is_binary(s) do
    case Jason.decode(s) do
      {:ok, v} -> v
      _ -> nil
    end
  end

  defp query(sql, params) do
    path = db_path()

    case File.exists?(path) do
      false ->
        {:error, :no_db}

      true ->
        with {:ok, conn} <- Sqlite3.open(path),
             {:ok, st} <- Sqlite3.prepare(conn, sql),
             :ok <- Sqlite3.bind(st, params),
             {:ok, rows} <- fetch_all(conn, st, []) do
          Sqlite3.release(conn, st)
          Sqlite3.close(conn)
          {:ok, rows}
        end
    end
  end

  defp fetch_all(conn, st, acc) do
    case Sqlite3.step(conn, st) do
      {:row, row} -> fetch_all(conn, st, [row | acc])
      :done -> {:ok, Enum.reverse(acc)}
      err -> err
    end
  end
end

File: bin/optimizer_runs.py
"""Persistent registry for async optimizer runs.

The user clicks "Optimize", may close the page, and come back later. We need
durable state for the optimizer job so the UI can show:
  - status (running, done, error, canceled)
  - log (live tail or final)
  - result (e.g. candidate prompt version, score metric)

The runner writes to this table; the LiveView polls/reads from it.

Schema lives in the central fabric DB so Python and Elixir share it.

Table: optimizer_run
  id           PK
  domain       TEXT  (the domain being optimized)
  target       TEXT  ("judge" | "generator")
  rubric       TEXT  (e.g. "pair-wheel-v2")
  prompt_name  TEXT  (langfuse prompt name)
  status       TEXT  ("running" | "done" | "error" | "canceled")
  exit_code    INTEGER
  log          TEXT  (accumulated stdout/stderr lines, newline-joined)
  result       TEXT  (JSON: {"candidate_version": N, "metric": 0.8, ...})
  started_at   TEXT
  finished_at  TEXT
"""
from __future__ import annotations
import json
import os
import sqlite3
from pathlib import Path

_SCHEMA = """
CREATE TABLE IF NOT EXISTS optimizer_run (
    id           INTEGER PRIMARY KEY AUTOINCREMENT,
    domain       TEXT    NOT NULL,
    target       TEXT    NOT NULL,
    rubric       TEXT,
    prompt_name  TEXT,
    status       TEXT    NOT NULL DEFAULT 'running',
    exit_code    INTEGER,
    log          TEXT    NOT NULL DEFAULT '',
    result       TEXT,
    started_at   TEXT    NOT NULL DEFAULT (datetime('now')),
    finished_at  TEXT
);
CREATE INDEX IF NOT EXISTS idx_optimizer_run_domain ON optimizer_run(domain);
CREATE INDEX IF NOT EXISTS idx_optimizer_run_status ON optimizer_run(status);
"""

def _db_path() -> Path:
    home = Path(os.environ.get("DATA_TOURNAMENTS_HOME", "/tmp/data-tournaments"))
    return home / "judgements.db"

def _connect() -> sqlite3.Connection:
    conn = sqlite3.connect(_db_path())
    conn.row_factory = sqlite3.Row
    conn.execute("PRAGMA busy_timeout = 5000")
    return conn

def init() -> None:
    """Create the optimizer_run table if missing. Idempotent."""
    with _connect() as conn:
        conn.executescript(_SCHEMA)

def start(*, domain: str, target: str, rubric: str | None = None,
          prompt_name: str | None = None) -> int:
    """Create a new running optimizer-run row, return its id."""
    init()
    with _connect() as conn:
        cur = conn.execute(
            "INSERT INTO optimizer_run(domain, target, rubric, prompt_name) "
            "VALUES (?, ?, ?, ?)",
            (domain, target, rubric, prompt_name),
        )
        return cur.lastrowid

def append_log(run_id: int, line: str) -> None:
    """Append one line to the run's log buffer."""
    with _connect() as conn:
        conn.execute(
            "UPDATE optimizer_run SET log = log || ? WHERE id = ?",
            (line + "\n", run_id),
        )

def finish(run_id: int, *, status: str, exit_code: int | None = None,
           result: dict | None = None) -> None:
    """Mark the run finished. status ∈ {done, error, canceled}."""
    if exit_code is None:
        exit_code = 0 if status == "done" else 1
    with _connect() as conn:
        conn.execute(
            "UPDATE optimizer_run "
            "SET status = ?, exit_code = ?, result = ?, "
            "    finished_at = datetime('now') "
            "WHERE id = ?",
            (status, exit_code, json.dumps(result) if result is not None else None, run_id),
        )

def get(run_id: int) -> dict | None:
    with _connect() as conn:
        row = conn.execute(
            "SELECT * FROM optimizer_run WHERE id = ?", (run_id,)
        ).fetchone()
        return _row_to_dict(row)

def list_for_domain(domain: str, *, limit: int = 50) -> list[dict]:
    with _connect() as conn:
        rows = conn.execute(
            "SELECT * FROM optimizer_run WHERE domain = ? "
            "ORDER BY id DESC LIMIT ?",
            (domain, limit),
        ).fetchall()
        return [_row_to_dict(r) for r in rows]

def latest(*, domain: str, target: str) -> dict | None:
    with _connect() as conn:
        row = conn.execute(
            "SELECT * FROM optimizer_run WHERE domain = ? AND target = ? "
            "ORDER BY id DESC LIMIT 1",
            (domain, target),
        ).fetchone()
        return _row_to_dict(row)

def _row_to_dict(row) -> dict | None:
    if row is None:
        return None
    d = dict(row)
    if d.get("result"):
        try:
            d["result"] = json.loads(d["result"])
        except (TypeError, ValueError):
            pass
    return d
